com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs

```diff
@@ -0,0 +1,411 @@
+        {
+            get { return previewCaret; }
+            set
+            {
+                if (previewCaret != value)
+                {
+                    previewCaret = value;
+                    InvalidateLayout();
+                }
+            }
+        }
+
+        private string text = string.Empty;
+
+        /// <summary>
+        /// The text to display in the preview.
+        /// </summary>
+        public string Text
+        {
+            get { return text; }
+            set
+            {
+                if (value != text)
+                {
+                    text = value;
+
+                    if (PreviewText != null)
+                    {
+                        PreviewText.text = text;
+                    }
+
+                    CaretIndex = Mathf.Clamp(CaretIndex, 0, Text.Length);
+                    InvalidateLayout();
+                }
+            }
+        }
+
+        private int caretIndex = 0;
+
+        /// <summary>
+        /// Where the caret lies within the text.
+        /// </summary>
+        public int CaretIndex
+        {
+            get { return caretIndex; }
+            set
+            {
+                value = Mathf.Clamp(value, 0, Text.Length);
+                if (value != caretIndex)
+                {
+                    caretIndex = value;
+                    InvalidateLayout();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Utility method which can be used to toggle if solvers update.
+        /// </summary>
+        public void ToggleSolvers()
+        {
+#if MRTK_SPATIAL_PRESENT
+            if (TryGetComponent(out SolverHandler solverHandler))
+            {
+                solverHandler.UpdateSolvers = !solverHandler.UpdateSolvers;
+
+                if (solverHandler.UpdateSolvers)
+                {
+                    ApplyShellSolverParameters();
+                }
+            }
+#endif
+        }
+
+        #region MonoBehaviour Implementation
+
+        private void OnEnable()
+        {
+            StartCoroutine(BlinkCaret());
+        }
+
+        private void Start()
+        {
+            ApplyShellSolverParameters();
+        }
+
+        private void LateUpdate()
+        {
+            if (layoutInvalidated)
+            {
+                layoutInvalidated = false;
+                UpdateLayout();
+            }
+
+        }
+
+        #endregion MonoBehaviour Implementation
+
+        /// <summary>
+        /// Force the <see cref="PreviewText"/> label and <see cref="PreviewCaret"/> transform to update their positions on the next update.
+        /// </summary>
+        public void InvalidateLayout()
+        {
+            layoutInvalidated = true;
+        }
+
+
+        private void UpdateLayout()
+        {
+            if (previewCaret == null || PreviewText == null)
+            {
+                return;
+            }
+
+            var textInfo = PreviewText.GetTextInfo(Text);
```

Comment:
Per frame allocation.  Can be a field var?

Reply:
TextInfo is a struct.  no heap allocation.  

This is needed unfortunately. TextInfo needs to be filled in here to start layout.  UpdatLayout() only happen with new keyboard input